Changes in module 'nf-core/gatk4/haplotypecaller'
--- modules/nf-core/gatk4/haplotypecaller/main.nf
+++ modules/nf-core/gatk4/haplotypecaller/main.nf
@@ -1,6 +1,6 @@
 process GATK4_HAPLOTYPECALLER {
     tag "$meta.id"
-    label 'process_medium'
+    label 'process_low'
 
     conda "${moduleDir}/environment.yml"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
@@ -44,7 +44,6 @@
         --input $input \\
         --output ${prefix}.vcf.gz \\
         --reference $fasta \\
-        --native-pair-hmm-threads ${task.cpus} \\
         $dbsnp_command \\
         $interval_command \\
         $dragstr_command \\

************************************************************
