Changes in module 'nf-core/gatk4/haplotypecaller'
--- modules/nf-core/gatk4/haplotypecaller/main.nf
+++ modules/nf-core/gatk4/haplotypecaller/main.nf
@@ -12,8 +12,8 @@
     tuple val(meta2), path(fasta)
     tuple val(meta3), path(fai)
     tuple val(meta4), path(dict)
-    tuple val(meta5), path(dbsnp)
-    tuple val(meta6), path(dbsnp_tbi)
+    path(dbsnp)
+    path(dbsnp_tbi)
 
     output:
     tuple val(meta), path("*.vcf.gz")       , emit: vcf

************************************************************
