Changes in module 'nf-core/somalier/relate'
--- modules/nf-core/somalier/relate/main.nf
+++ modules/nf-core/somalier/relate/main.nf
@@ -16,6 +16,7 @@
     tuple val(meta), path("*.html"),          emit: html
     tuple val(meta), path("*.pairs.tsv"),     emit: pairs_tsv
     tuple val(meta), path("*.samples.tsv"),   emit: samples_tsv
+    tuple val(meta), path("*.ped"),           emit: ped
     path "versions.yml",                      emit: versions
 
     when:
@@ -36,6 +37,8 @@
         ${sample_groups_command} \\
         ${ped_command}
 
+    cut -f1-6 ${prefix}.samples.tsv | sed -e 's/-9/0/' > ${prefix}_somalier.ped
+
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":
         somalier: \$(echo \$(somalier 2>&1) | sed 's/^.*somalier version: //; s/Commands:.*\$//')
@@ -49,6 +52,7 @@
     touch ${prefix}.html
     touch ${prefix}.pairs.tsv
     touch ${prefix}.samples.tsv
+    touch ${prefix}_somalier.ped
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
