Changes in module 'nf-core/ensemblvep/vep'
--- modules/nf-core/ensemblvep/vep/main.nf
+++ modules/nf-core/ensemblvep/vep/main.nf
@@ -1,6 +1,6 @@
 process ENSEMBLVEP_VEP {
     tag "$meta.id"
-    label 'process_medium'
+    label 'process_single'
 
     conda "bioconda::ensembl-vep=108.2"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
@@ -18,6 +18,7 @@
 
     output:
     tuple val(meta), path("*.vcf.gz")  , optional:true, emit: vcf
+    tuple val(meta), path("*.tbi")     , optional:true, emit: tbi
     tuple val(meta), path("*.tab.gz")  , optional:true, emit: tab
     tuple val(meta), path("*.json.gz") , optional:true, emit: json
     path "*.summary.html"              , emit: report
@@ -48,6 +49,7 @@
         --fork $task.cpus \\
         --stats_file ${prefix}.summary.html \\
 
+    tabix ${prefix}.${file_extension}.gz
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":
@@ -59,6 +61,7 @@
     def prefix = task.ext.prefix ?: "${meta.id}"
     """
     touch ${prefix}.vcf.gz
+    touch ${prefix}.tbi
     touch ${prefix}.tab.gz
     touch ${prefix}.json.gz
     touch ${prefix}.summary.html

************************************************************
