Changes in module 'nf-core/ensemblvep/vep'
--- modules/nf-core/ensemblvep/vep/main.nf
+++ modules/nf-core/ensemblvep/vep/main.nf
@@ -2,7 +2,7 @@
     tag "$meta.id"
     label 'process_medium'
 
-    conda "bioconda::ensembl-vep=108.2"
+    conda "bioconda::ensembl-vep=105.0"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
         'https://depot.galaxyproject.org/singularity/ensembl-vep:108.2--pl5321h4a94de4_0' :
         'quay.io/biocontainers/ensembl-vep:108.2--pl5321h4a94de4_0' }"

************************************************************
