Changes in module 'nf-core/ensemblvep/vep'
--- modules/nf-core/ensemblvep/vep/main.nf
+++ modules/nf-core/ensemblvep/vep/main.nf
@@ -2,10 +2,10 @@
     tag "$meta.id"
     label 'process_medium'
 
-    conda "bioconda::ensembl-vep=109.3"
+    conda "${moduleDir}/environment.yml"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
-        'https://depot.galaxyproject.org/singularity/ensembl-vep:109.3--pl5321h2a3209d_1' :
-        'biocontainers/ensembl-vep:109.3--pl5321h2a3209d_1' }"
+        'https://depot.galaxyproject.org/singularity/ensembl-vep:105.0--pl5321h4a94de4_1' :
+        'biocontainers/ensembl-vep:105.0--pl5321h4a94de4_1' }"
 
     input:
     tuple val(meta), path(vcf), path(custom_extra_files)
@@ -48,7 +48,6 @@
         --fork $task.cpus \\
         --stats_file ${prefix}.summary.html \\
 
-
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":
         ensemblvep: \$( echo \$(vep --help 2>&1) | sed 's/^.*Versions:.*ensembl-vep : //;s/ .*\$//')

--- /dev/null
+++ modules/nf-core/ensemblvep/vep/environment.yml
@@ -0,0 +1,7 @@
+name: ensemblvep_download
+channels:
+  - conda-forge
+  - bioconda
+  - defaults
+dependencies:
+  - bioconda::ensembl-vep=105.0

************************************************************
