Changes in module 'nf-core/ensemblvep'
--- modules/nf-core/ensemblvep/environment.yml
+++ modules/nf-core/ensemblvep/environment.yml
@@ -1,10 +1,10 @@
 # You can use this file to create a conda environment for this module:
 #   conda env create -f environment.yml
-name: nf-core-vep-106.1
+name: nf-core-vep-105.0
 channels:
   - conda-forge
   - bioconda
   - defaults
 
 dependencies:
-  - bioconda::ensembl-vep=106.1
+  - bioconda::ensembl-vep=105.0

--- modules/nf-core/ensemblvep/Dockerfile
+++ modules/nf-core/ensemblvep/Dockerfile
@@ -11,8 +11,8 @@
 # Setup default ARG variables
 ARG GENOME=GRCh38
 ARG SPECIES=homo_sapiens
-ARG VEP_CACHE_VERSION=106
-ARG VEP_VERSION=106.1
+ARG VEP_CACHE_VERSION=105
+ARG VEP_VERSION=105.0
 
 # Add conda installation dir to PATH (instead of doing 'conda activate')
 ENV PATH /opt/conda/envs/nf-core-vep-${VEP_VERSION}/bin:$PATH

--- modules/nf-core/ensemblvep/build.sh
+++ modules/nf-core/ensemblvep/build.sh
@@ -20,9 +20,9 @@
     docker push nfcore/vep:${VEP_VERSION}.${GENOME}
 }
 
-build_push "GRCh37"    "homo_sapiens"           "106" "106.1"
-build_push "GRCh38"    "homo_sapiens"           "106" "106.1"
-build_push "GRCm38"    "mus_musculus"           "102" "106.1"
-build_push "GRCm39"    "mus_musculus"           "106" "106.1"
-build_push "CanFam3.1" "canis_lupus_familiaris" "104" "106.1"
-build_push "WBcel235"  "caenorhabditis_elegans" "106" "106.1"
+build_push "GRCh37"    "homo_sapiens"           "105" "105.0"
+build_push "GRCh38"    "homo_sapiens"           "105" "105.0"
+build_push "GRCm38"    "mus_musculus"           "105" "105.0"
+build_push "GRCm39"    "mus_musculus"           "105" "105.0"
+build_push "CanFam3.1" "canis_lupus_familiaris" "105" "105.0"
+build_push "WBcel235"  "caenorhabditis_elegans" "105" "105.0"

--- modules/nf-core/ensemblvep/main.nf
+++ modules/nf-core/ensemblvep/main.nf
@@ -2,10 +2,10 @@
     tag "$meta.id"
     label 'process_medium'
 
-    conda (params.enable_conda ? "bioconda::ensembl-vep=106.1" : null)
+    conda (params.enable_conda ? "bioconda::ensembl-vep=105.0" : null)
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
-        'https://depot.galaxyproject.org/singularity/ensembl-vep:106.1--pl5321h4a94de4_0' :
-        'quay.io/biocontainers/ensembl-vep:106.1--pl5321h4a94de4_0' }"
+        'https://depot.galaxyproject.org/singularity/ensembl-vep:105.0--pl5321h4a94de4_1' :
+        'quay.io/biocontainers/ensembl-vep:105.0--pl5321h4a94de4_1' }"
 
     input:
     tuple val(meta), path(vcf)
@@ -17,12 +17,9 @@
     path  extra_files
 
     output:
-    tuple val(meta), path("*.ann.vcf")     , optional:true, emit: vcf
-    tuple val(meta), path("*.ann.tab")     , optional:true, emit: tab
-    tuple val(meta), path("*.ann.json")    , optional:true, emit: json
-    tuple val(meta), path("*.ann.vcf.gz")  , optional:true, emit: vcf_gz
-    tuple val(meta), path("*.ann.tab.gz")  , optional:true, emit: tab_gz
-    tuple val(meta), path("*.ann.json.gz") , optional:true, emit: json_gz
+    tuple val(meta), path("*.ann.{vcf,vcf.gz}")     , optional:true, emit: vcf
+    tuple val(meta), path("*.ann.{tab,tab.gz}")     , optional:true, emit: tab
+    tuple val(meta), path("*.ann.{json,json.gz}")   , optional:true, emit: json
     path "*.summary.html"                  , emit: report
     path "versions.yml"                    , emit: versions
 

************************************************************
