Changes in module 'nf-core/bedtools/intersect'
--- modules/nf-core/bedtools/intersect/main.nf
+++ modules/nf-core/bedtools/intersect/main.nf
@@ -8,7 +8,7 @@
         'quay.io/biocontainers/bedtools:2.30.0--hc088bd4_0' }"
 
     input:
-    tuple val(meta), path(intervals1), path(intervals2)
+    tuple val(meta), path(intervals1), path(intervals2, stageAs:"?/*")
     val extension
 
     output:

************************************************************
