Changes in module 'nf-core/bedtools/intersect'
--- modules/nf-core/bedtools/intersect/main.nf
+++ modules/nf-core/bedtools/intersect/main.nf
@@ -44,7 +44,7 @@
 
     stub:
     def prefix = task.ext.prefix ?: "${meta.id}"
-    def extension = tast.ext.suffix ?: "bed"
+    extension = task.ext.suffix ?: "bed"
     if ("$intervals1" == "${prefix}.${extension}" ||
         "$intervals2" == "${prefix}.${extension}")
         error "Input and output names are the same, use \"task.ext.prefix\" to disambiguate!"

************************************************************
