Changes in module 'nf-core/vardictjava'
--- modules/nf-core/vardictjava/main.nf
+++ modules/nf-core/vardictjava/main.nf
@@ -39,6 +39,7 @@
     | ${filter} \\
     | ${convert_to_vcf} \\
         ${args2} \\
+    | sed -e 's/##FORMAT=<ID=AF,Number=A/##FORMAT=<ID=AF,Number=1/' \\
     > ${prefix}.vcf
 
     cat <<-END_VERSIONS > versions.yml

************************************************************
