Changes in module 'nf-core/gatk4/genomicsdbimport'
--- modules/nf-core/gatk4/genomicsdbimport/main.nf
+++ modules/nf-core/gatk4/genomicsdbimport/main.nf
@@ -2,7 +2,7 @@
     tag "$meta.id"
     label 'process_low'
 
-    conda "bioconda::gatk4=4.3.0.0"
+    conda (params.enable_conda ? "bioconda::gatk4=4.3.0.0" : null)
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
         'https://depot.galaxyproject.org/singularity/gatk4:4.3.0.0--py36hdfd78af_0':
         'quay.io/biocontainers/gatk4:4.3.0.0--py36hdfd78af_0' }"
@@ -70,6 +70,7 @@
 
     genomicsdb_command = "--genomicsdb-workspace-path ${prefix}"
     interval_command = interval_file ? "--intervals ${interval_file}" : "--intervals ${interval_value}"
+    updated_db = []
 
     // settings changed for running get intervals list mode if run_intlist is true
     if (run_intlist) {

************************************************************
