Changes in module 'nf-core/gatk4/genomicsdbimport'
--- modules/nf-core/gatk4/genomicsdbimport/main.nf
+++ modules/nf-core/gatk4/genomicsdbimport/main.nf
@@ -1,6 +1,6 @@
 process GATK4_GENOMICSDBIMPORT {
     tag "$meta.id"
-    label 'process_medium'
+    label 'process_single'
 
     conda "bioconda::gatk4=4.4.0.0"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?

************************************************************
