Changes in module 'nf-core/bedtools/split'
--- modules/nf-core/bedtools/split/main.nf
+++ modules/nf-core/bedtools/split/main.nf
@@ -9,6 +9,7 @@
 
     input:
     tuple val(meta), path(bed)
+    val(number_of_files)
 
     output:
     tuple val(meta), path("*.bed"), emit: beds
@@ -26,7 +27,8 @@
         split \\
         $args \\
         -i $bed \\
-        -p $prefix
+        -p $prefix \\
+        -n $number_of_files
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
